src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -100,37 +87,273 @@ Status RedisStoreClient::AsyncGet(const std::string &table_name, const std::stri
+    table_keys.push_back(GenTableKey(table_name, key));
+  }
+  return DeleteByKeys(table_keys, callback);
+}
+
 Status RedisStoreClient::AsyncDeleteByIndex(const std::string &table_name,
                                             const std::string &index_key,
                                             const StatusCallback &callback) {
-  RAY_CHECK(0) << "Not implemented! Will implement this function in next PR.";
+  std::string match_pattern = GenIndexTableMatchPattern(index_key, table_name);
+  auto scanner = std::make_shared<RedisScanner>(redis_client_, table_name, match_pattern);
+  auto on_done = [this, table_name, index_key, callback, scanner](
+                     const Status &status, const std::vector<std::string> &result) {
+    if (!result.empty()) {
+      std::vector<std::string> keys;
+      keys.reserve(result.size());
+      for (auto &item : result) {
+        keys.push_back(GetKeyFromIndexTableKey(item, table_name, index_key));
+      }
+      auto on_done = [this, result, callback](const Status &status) {
+        RAY_CHECK_OK(status);
+        // Delete index keys.
+        RAY_CHECK_OK(DeleteByKeys(result, callback));
+      };
+      RAY_CHECK_OK(AsyncBatchDelete(table_name, keys, callback));
+    } else {
+      callback(status);
+    }
+  };
+
+  return scanner->ScanKeys(on_done);
+}
+
+Status RedisStoreClient::DoPut(const std::string &key, const std::string &data,
+                               const StatusCallback &callback) {
+  std::vector<std::string> args = {"SET", key, data};
+  RedisCallback write_callback = nullptr;
+  if (callback) {
+    write_callback = [callback](const std::shared_ptr<CallbackReply> &reply) {
+      auto status = reply->ReadAsStatus();
+      callback(status);
+    };
+  }
+
+  auto shard_context = redis_client_->GetShardContext(key);
+  return shard_context->RunArgvAsync(args, write_callback);
+}
+
+Status RedisStoreClient::DeleteByKeys(const std::vector<std::string> &keys,
+                                      const StatusCallback &callback) {
+  std::unordered_map<RedisContext *, std::vector<std::string>> shards;
+  for (auto &key : keys) {
+    auto shard_context = redis_client_->GetShardContext(key).get();
+    auto it = shards.find(shard_context);
+    if (it == shards.end()) {
+      shards[shard_context].push_back("DEL");
+      shards[shard_context].push_back(key);
+    } else {
+      it->second.push_back(key);
+    }
+  }
+
+  auto finished_count = std::make_shared<int>(0);
+  int size = shards.size();
+  for (auto &item : shards) {
+    auto delete_callback = [finished_count, size,
+                            callback](const std::shared_ptr<CallbackReply> &reply) {
+      ++(*finished_count);
+      if (*finished_count == size) {
+        callback(Status::OK());
+      }
+    };
+    RAY_CHECK_OK(item.first->RunArgvAsync(item.second, delete_callback));
+  }
   return Status::OK();
 }
 
+std::string RedisStoreClient::GenTableKey(const std::string &table_name,
```

Comment:
```suggestion
std::string RedisStoreClient::GenRedisKey(const std::string &table_name,
```
There are a couple of keys: `table key`, `full key`, `key`, which is confusing. 
I think the first 2 are the same. we can just call it `redis key`.

Reply:
fixed